File: src/connectors/crt_sh.js
import { BaseConnector } from './base.js';
import https from 'https';

export class CrtShConnector extends BaseConnector {
    constructor(config, state, telemetry) {
        super(config, state, telemetry);
        this.name = 'crt.sh';
    }

    async run(investigationId, phaseId, inputType, inputValue) {
        const runId = this.state.recordConnectorRun(investigationId, phaseId, this.name, inputType, inputValue);
        this.telemetry.connectorStart(investigationId, this.name, phaseId, { type: inputType, value: inputValue });

        try {
            const data = await this._fetch(`https://crt.sh/?q=%25.${encodeURIComponent(inputValue)}&output=json`);
            if (data.trimStart().startsWith('<')) throw new Error('crt.sh returned HTML (rate limited or down)');
            const records = JSON.parse(data);
            const entities = [];
            const seen = new Set();

            for (const cert of records) {
                const names = (cert.name_value || '').split('\n');
                for (const name of names) {
                    const clean = name.trim().replace(/^\*\./, '');
                    if (!clean || seen.has(clean)) continue;
                    seen.add(clean);

                    entities.push({
                        type: 'subdomain',
                        data: {
                            name: clean,
                            parent_domain: inputValue,
                            source: 'certificate_transparency'
                        },
                        confidence: 0.9,
                        asKnown: { type: 'domain', value: clean }
                    });
                }

                if (cert.issuer_name && !seen.has('cert:' + cert.serial_number)) {
                    seen.add('cert:' + cert.serial_number);
                    entities.push({
                        type: 'certificate',
                        data: {
                            subject: cert.common_name,
                            issuer: cert.issuer_name,
                            not_before: cert.not_before,
                            not_after: cert.not_after,
                            serial: cert.serial_number,
                            san: cert.name_value
                        },
                        confidence: 0.95
                    });
                }
            }

            let newCount = 0;
            for (const entity of entities) {
                const added = this.state.addEntity(investigationId, entity.type, entity.data, entity.confidence || 0.5);
                if (added.new) newCount++;
                this.telemetry.entityDiscovered(investigationId, entity.type, added.new, this.name);
                if (entity.asKnown) {
                    this.state.addKnown(investigationId, entity.asKnown.type, entity.asKnown.value, this.name, entity.confidence || 0.5);
                }
            }

            this.state.completeConnectorRun(runId, 'completed', newCount, null, null, 0);
            this.telemetry.connectorEnd(investigationId, this.name, phaseId, {
                status: 'completed', entitiesFound: newCount, input: { type: inputType, value: inputValue }
            });
            return { status: 'completed', entities, newCount };
        } catch (e) {
            this.state.completeConnectorRun(runId, 'failed', 0, null, e.message, 1);
            this.telemetry.connectorEnd(investigationId, this.name, phaseId, {
                status: 'failed', entitiesFound: 0, input: { type: inputType, value: inputValue }
            });
            return { status: 'failed', error: e.message };
        }
    }

    _fetch(url) {
        return new Promise((resolve, reject) => {
            const req = https.get(url, { headers: { 'User-Agent': 'cyclops-osint' }, timeout: 30000 }, res => {
                if (res.statusCode === 429) { reject(new Error('crt.sh returned HTML (rate limited or down)')); res.resume(); return; }
                if (res.statusCode >= 400) { reject(new Error(`crt.sh HTTP ${res.statusCode}`)); res.resume(); return; }
                let data = '';
                res.on('data', chunk => { data += chunk; if (data.length > 5e6) { req.destroy(); reject(new Error('response too large')); } });
                res.on('end', () => resolve(data));
            });
            req.on('error', reject);
            req.on('timeout', () => { req.destroy(); reject(new Error('crt.sh timeout')); });
        });
    }

    async healthCheck() { return { ok: true, version: 'native' }; }
}

File: cyclops/connectors/crt_sh.py
import aiohttp
from cyclops.connectors.base import BaseConnector


class CrtShConnector(BaseConnector):
    def __init__(self, config, state, telemetry):
        super().__init__(config, state, telemetry)
        self.name = "crt.sh"

    async def run(self, investigation_id, phase_id, input_type, input_value):
        run_id = self.state.record_connector_run(investigation_id, phase_id, self.name, input_type, input_value)
        self.telemetry.connector_start(investigation_id, self.name, phase_id, {"type": input_type, "value": input_value})

        try:
            entities = []
            url = f"https://crt.sh/?q=%.{input_value}&output=json"
            async with aiohttp.ClientSession() as session:
                async with session.get(url, timeout=aiohttp.ClientTimeout(total=30)) as resp:
                    if resp.status != 200:
                        raise Exception(f"crt.sh HTTP {resp.status}")
                    data = await resp.json(content_type=None)

            seen = set()
            for entry in (data or []):
                name = entry.get("name_value", "")
                for sub in name.split("\n"):
                    sub = sub.strip().lower().lstrip("*.")
                    if sub and sub != input_value and sub not in seen:
                        seen.add(sub)
                        entities.append({
                            "type": "subdomain",
                            "data": {"name": sub, "parent_domain": input_value, "source": "crt_sh"},
                            "confidence": 0.9,
                            "as_known": {"type": "domain", "value": sub},
                        })

                issuer = entry.get("issuer_name", "")
                if issuer:
                    entities.append({
                        "type": "certificate",
                        "data": {"subject": entry.get("common_name"), "issuer": issuer, "not_before": entry.get("not_before"), "not_after": entry.get("not_after"), "source": "crt_sh"},
                        "confidence": 0.85,
                    })

            seen_certs = set()
            deduped = []
            for e in entities:
                if e["type"] == "certificate":
                    key = e["data"].get("subject", "")
                    if key in seen_certs:
                        continue
                    seen_certs.add(key)
                deduped.append(e)

            new_count = 0
            for entity in deduped:
                added = self.state.add_entity(investigation_id, entity["type"], entity["data"], entity.get("confidence", 0.5))
                if added["new"]:
                    new_count += 1
                self.telemetry.entity_discovered(investigation_id, entity["type"], added["new"], self.name)
                if "as_known" in entity:
                    self.state.add_known(investigation_id, entity["as_known"]["type"], entity["as_known"]["value"], self.name)

            self.state.complete_connector_run(run_id, "completed", new_count)
            self.telemetry.connector_end(investigation_id, self.name, phase_id, {"status": "completed", "entitiesFound": new_count, "input": {"type": input_type, "value": input_value}})
            return {"status": "completed", "entities": deduped, "newCount": new_count}
        except Exception as e:
            self.state.complete_connector_run(run_id, "failed", 0, error=str(e))
            self.telemetry.connector_end(investigation_id, self.name, phase_id, {"status": "failed", "input": {"type": input_type, "value": input_value}})
            return {"status": "failed", "error": str(e)}

    async def health_check(self):
        return {"ok": True, "version": "free"}
